src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,359 @@
+            throws AuthenticationException {
+        Logger.v(TAG, "Try to get new access token with the found refresh token.", 
+                mAuthRequest.getLogInfo(), null);
+        
+        // @note: The original intention for checking network connection before sending RT request is to avoid RT removal
+        // from cache if user turned on airplane mode or similar case. However, the updated cache logic only remove RT when 
+        // receiving oauth2 error invalid_grant from server. 
+        // The only reason to keep it here is in case there is caller taking dependency on the below error code. 
+        final DefaultConnectionService connectionService = new DefaultConnectionService(mContext);
+        if (!connectionService.isConnectionAvailable()) {
+            AuthenticationException authenticationException = new AuthenticationException(
+                    ADALError.DEVICE_CONNECTION_IS_NOT_AVAILABLE,
+                    "Connection is not available to refresh token");
+            Logger.w(TAG, "Connection is not available to refresh token", mAuthRequest.getLogInfo(),
+                    ADALError.DEVICE_CONNECTION_IS_NOT_AVAILABLE);
+            
+            throw authenticationException;
+        }
+        
+        final AuthenticationResult result;
+        try {
+            final JWSBuilder jwsBuilder = new JWSBuilder();
+            final Oauth2 oauthRequest = new Oauth2(mAuthRequest, mWebRequestHandler, jwsBuilder);
+            result = oauthRequest.refreshToken(refreshToken);
+            if (result != null && StringExtensions.IsNullOrBlank(result.getRefreshToken())) {
+                Logger.v(TAG, "Refresh token is not returned or empty");
+                result.setRefreshToken(refreshToken);
+            }
+        } catch (final IOException | AuthenticationException exc) {
+            // Server side error or similar
+            Logger.e(TAG, "Error in refresh token for request:" + mAuthRequest.getLogInfo(),
+                    ExceptionExtensions.getExceptionMessage(exc), ADALError.AUTH_FAILED_NO_TOKEN,
+                    exc);
+
+            AuthenticationException authException = new AuthenticationException(
+                    ADALError.AUTH_FAILED_NO_TOKEN, ExceptionExtensions.getExceptionMessage(exc),
+                    exc);
+            throw authException;
+        }
+
+        return result;
+    }
+    
+    /**
+     * For testing purpose, inject the dependency. 
+     */
+    void setWebRequestHandler(final IWebRequestHandler webRequestHandler) {
+        this.mWebRequestHandler = webRequestHandler;
+    }
+    
+    /**
+     * Attempt to get new access token with regular RT. 
+     */
+    private AuthenticationResult tryRT() throws AuthenticationException {
+        final String user = getUser();
+        final TokenCacheItem regularRTItem = mTokenCacheAccessor.getRegularTokenCacheItemWithAT(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), user);
```

Comment:
Done. 

---
In reply to: [62920466](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r62920466) [](ancestors = 62920466)